netty/src/test/java/io/grpc/transport/netty/NettyClientHandlerTest.java

```diff
@@ -135,10 +143,38 @@ public void setUp() throws Exception {
   }
 
   @Test
-  public void createStreamShouldSucceed() throws Exception {
-    handler.write(ctx, new CreateStreamCommand(grpcHeaders, stream),
-        promise);
+  public void cancelBufferedStreamShouldChangeClientStreamStatus() throws Exception {
+    // Force the stream to be buffered.
+    setMaxConcurrentStreams(0);
+    // Create a new stream with id 3.
+    ChannelPromise createPromise =
+            new DefaultChannelPromise(channel, ImmediateEventExecutor.INSTANCE);
+    handler.write(ctx, new CreateStreamCommand(grpcHeaders, stream), createPromise);
+    verify(stream).id(eq(3));
+    when(stream.id()).thenReturn(3);
+    // Cancel the stream.
+    handler.write(ctx, new CancelStreamCommand(stream), promise);
+
     verify(promise).setSuccess();
+
+    assertTrue(createPromise.isDone());
+    assertFalse(createPromise.isSuccess());
+    ArgumentCaptor<Status> statusCaptor = ArgumentCaptor.forClass(Status.class);
+    verify(stream, times(2)).transportReportStatus(statusCaptor.capture(), eq(true),
+                                                   any(Metadata.Trailers.class));
+    List<Status> statuses = statusCaptor.getAllValues();
+    Status canceled = statuses.get(0);
+    Status internal = statuses.get(1);
+
+    assertEquals(Status.CANCELLED, canceled);
+    assertEquals(Status.INTERNAL.getCode(), internal.getCode());
+    assertThat(internal.getCause(), instanceOf(Http2Exception.class));
```

Comment:
:-/ So cancelling the stream causes an exception, but only if it was buffered. I really think we should just succeed the promises in writeRstStream() instead.

It can be useful to know that the data wasn't sent, but by failing the stream we conflate an error writing with being guaranteed that the server didn't process the stream. It would be better for most of the code to just believe it was sent to the server and immediately cancelled, which semantically is fully equivalent to what did happen, than believe something went wrong.